src/BaseUriPlugin.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+class BaseUriPlugin implements Plugin
+{
+    /**
+     * @var UriInterface
+     */
+    private $baseUri;
+
+    /**
+     * @param UriInterface $baseUri
+     */
+    public function __construct(UriInterface $baseUri)
+    {
+        $this->baseUri = $baseUri;
+    }
+
+    /**
+     * Handle the request and return the response coming from the next callable.
+     *
+     * @param RequestInterface $request
+     * @param callable         $next    Next middleware in the chain, the request is passed as the first argument
+     * @param callable         $first   First middleware in the chain, used to to restart a request
+     *
+     * @return Promise
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $uriString = $request->getUri()->__toString();
+        if (!strpos($uriString, '://')) {
```

Comment:
Hm, I've looked at Guzzle's implementation and copied their functionality. Which means that if we have the following: 

| Base path | Input | Output |
| ------------- | ----- | ------ |
| http://foo.com/api/ | http://baz.com/biz | http://baz.com/biz 
| http://foo.com/api/ | foo | http://foo.com/api/foo
| http://foo.com/api/ | /foo | http://foo.com/foo

Do I understand you correctly when you suggest the base path should always replace the available parts of the input?
It is a great idea but I can see a problem with that. How do you know what to replace?

| Base path | Input | Output |
| ------------- | ----- | ------ |
| http://foo.com/api/ | http://baz.com/biz/bar | http://foo.com/api/biz/bar, http://foo.com/api/bar or http://foo.com/api/?

I believe that the one that writes the api client should use this plugin to allow users to replace the base path with whatever. 

I would suggest a ReplacePath plugin to implement your functionality.